### PR TITLE
Chatbot Language Modelling Training Blueprint & Batch Predict Blueprint - For Review

### PR DESCRIPTION
**Pull request by yidinghabana**
Please review the following six libraries for two new chatbot language modelling blueprints:

ft-s3-connector
ft-fine-tune-bert
ft-fine-tune-distilgpt2
ft-fine-tune-gpt2
ft-fine-tune-compare
ft-fine-tune-inference-batch

### CODE REVIEW

File: ft-fine-tune-bert/fine-tune-unittests.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import transformers
+from transformers import PreTrainedTokenizer, LineByLineTextDataset, TextDataset, AutoTokenizer
+
+print(os.getcwd())
+
+class TestFT(unittest.TestCase):
+    def setUp(self):
+        """Overrides setUp from unittest to create a toy pandas dataframe for testing"""
+        # Here anomalies and non-anomalies are represented by -1 and 1 respectively.
+        self.train_data = [
```
Comment (orionsBeltWest) [inline ft-fine-tune-bert/fine-tune-unittests.py]: probably should dump the train_data into external file to be read from? So it could be swapped as needed. 

yaml or json comes to mind.

File: ft-fine-tune-bert/fine-tune-unittests.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        """Checks if input file existing"""
+        file = pl.Path("./6_genre_clean_training_data_small.txt")
+        self.assertIsFile(file)
+
+    # Test 2
+    def test_data_paths(self):
+        """Checks if input file existing"""
+        file = pl.Path("./6_genre_eval_data_small.txt")
+        self.assertIsFile(file)
+
```
Comment (orionsBeltWest) [inline ft-fine-tune-bert/fine-tune-unittests.py:31]: self.assertIsFile(file) and self.assertDirExist(file) raise exceptions. 

Hence you have to use the unittest exception testing methods. See https://github.com/cnvrg/unittest-framework/blob/main/connectorUnittest/test_connector_factory.py for example.

File: ft-fine-tune-bert/fine-tune-unittests.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            eval_data_file='6_genre_eval_data_small_unittests.txt',
+            line_by_line=True,
+            block_size=256,
+
+            overwrite_cache=True,
+        )
+        tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
+        returned_lists_trainy = get_dataset(data_args, tokenizer=tokenizer)
+        returned_lists_evaly = get_dataset(data_args, tokenizer=tokenizer, evaluate=True)
+        self.assertIsInstance(returned_lists_trainy, transformers.data.datasets.language_modeling.LineByLineTextDataset)
```
Comment (orionsBeltWest) [inline ft-fine-tune-bert/fine-tune-unittests.py:63]: Return type testing is not enough. 

Need to add test cases to actual training / fine tune performance.

File: ft-fine-tune-compare/compare-unittests.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    def assertIsFile(self, path):
+        if not pl.Path(path).resolve().is_file():
+            raise AssertionError("File does not exist: %s" % str(path))
+    
+    # Test 1
+    def test_data_paths(self):
+        """Checks if input file existing"""
+        file = pl.Path("eval_results_lm_1.txt")
+        self.assertIsFile(file)
```
Comment (orionsBeltWest) [inline ft-fine-tune-compare/compare-unittests.py:69]: same issue with asserIsFile and dir. Need to test for exceptions.

File: ft-fine-tune-gpt2/fine-tune-unittests.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                f"Output directory ({path}) already exists and is not empty. Use --overwrite_output_dir to overcome."
+            )
+
+    # Test 1
+    def test_data_paths(self):
+        """Checks if input file existing"""
+        file = pl.Path("./6_genre_clean_training_data_small.txt")
+        self.assertIsFile(file)
+
+    # Test 2
```
Comment (orionsBeltWest) [inline ft-fine-tune-gpt2/fine-tune-unittests.py:27]: same issue as bert unit test

File: ft-fine-tune-distilgpt2/fine-tune-unittests.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
+import pandas as pd
+import unittest
+from finetune import DataTrainingArguments, get_dataset
+import os
+import pathlib as pl
+import transformers
+from transformers import PreTrainedTokenizer, LineByLineTextDataset, TextDataset, AutoTokenizer
+
+print(os.getcwd())
```
Comment (orionsBeltWest) [inline ft-fine-tune-distilgpt2/fine-tune-unittests.py]: same issues as bert unittest

File: ft-fine-tune-inference-batch/inference-batch-unittests.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            [
+                "<BOS> <drama> Liz is a Los"
+            ],
+         ]
+
+        self.train_data_list = []
+        for row in self.train_data:
+            for col in row: 
+                self.train_data_list.append(str(col))
+        # print("===self.train_data_list===", self.train_data_list)
```
Comment (orionsBeltWest) [inline ft-fine-tune-inference-batch/inference-batch-unittests.py:39]: same issues with compare unittest

**Review by orionsBeltWest — CHANGES_REQUESTED** [review]
All these using the same unittest.  Try using classes and inheritance to eliminated code duplicate and minimize effort in fixing issues common to all. 

For exception you need to test using:
``` python:
        with self.assertRaises(someExceptionError):
            some calss of function call
```

Consider the application metric: 
Testing file's location return type is fine. But this deep learning app. So, we need test did we get an acceptable result?  
So I would except the main (most important) test cases focusing on the success metric

File: ft-fine-tune-bert/fine-tune-unittests.py (hunk 1 above)
Comment (yidinghabana) [reply ft-fine-tune-bert/fine-tune-unittests.py]: Add two external files: unittests_eval_data.txt and unittests_training_data.txt for Unittests to read data from.

File: ft-fine-tune-bert/fine-tune-unittests.py (hunk 3 above)
Comment (yidinghabana) [reply ft-fine-tune-bert/fine-tune-unittests.py:63]: Add unittests to check performance metrics perplexity.

**Comment by yidinghabana** [comment]
Add unittests to check performance metrics perplexity for all training libraries. 
Add two external files: unittests_eval_data.txt and unittests_training_data.txt for Unittests to read data from.

File: ft-fine-tune-bert/fine-tune-unittests.py [hunk 8]
```diff
@@ -1,55 +1,72 @@
...
+    #     with open("6_genre_clean_training_data_small_unittests.txt", "w") as output:
+    #         for row in self.train_data:
+    #             for col in row: 
+    #                 output.write(str(col) + '\n')
 
-        with open("6_genre_eval_data_small_unittests.txt", "w") as output:
-            for row in self.eval_data:
-                for col in row:
-                    output.write(str(col) + '\n')
+    #     with open("6_genre_eval_data_small_unittests.txt", "w") as output:
```
Comment (orionsBeltWest) [inline ft-fine-tune-bert/fine-tune-unittests.py]: please remove/delete the commented code as its huge blocks and it confuses the person reading the code.

File: ft-fine-tune-bert/fine-tune-unittests.py [hunk 9]
```diff
@@ -1,55 +1,72 @@
...
+    #         for row in self.eval_data:
+    #             for col in row:
+    #                 output.write(str(col) + '\n')
 
 
     def assertIsFile(self, path):
```
Comment (orionsBeltWest) [inline ft-fine-tune-bert/fine-tune-unittests.py]: This function which you calling in your test cases is raising exceptions.   Hence you need to rewrite test 1 and 2. 

I mentioned before you should not raise exceptions in your testing code.  However, if your application under  test  throwing exceptions then you could test these.

File: ft-fine-tune-bert/fine-tune-unittests.py [hunk 10]
```diff
@@ -110,9 +127,110 @@ def test_return_type(self):
...
-            os.remove(data_args.train_data_file)
-            os.remove(data_args.eval_data_file)
+        # if data_args.train_data_file and data_args.eval_data_file:
+        #     os.remove(data_args.train_data_file)
+        #     os.remove(data_args.eval_data_file)
+
+    # Test 6
+    def test_return_metrics(self):
+        """Checks if the metrics peplexity meet target value."""
+        model_name = "bert-base-uncased"
```
Comment (orionsBeltWest) [inline ft-fine-tune-bert/fine-tune-unittests.py:71]: Too much code here.  This should be wrapped in a function in your application.

File: ft-fine-tune-bert/fine-tune-unittests.py [hunk 11]
```diff
@@ -110,9 +127,110 @@ def test_return_type(self):
...
+
+        # Evaluation
+        if training_args.do_eval:
+            logger.info("*** Evaluate ***")
+            eval_output = trainer.evaluate()
+            perplexity = math.exp(eval_output["eval_loss"])
+        message = "Perplexity is not less that target value."
+        print("===Perplexity===", perplexity)
+        self.assertLess(perplexity, 220, message)
+        if output_model_path+model_name:
```
Comment (orionsBeltWest) [inline ft-fine-tune-bert/fine-tune-unittests.py:94]: add this line to cleanup part of the class

**Review by orionsBeltWest — CHANGES_REQUESTED** [review]
Added my comments in ft-fine-tune-bert.   But these also apply to ft-fine-tune-distilgpt2 and ft-fine-tune-gpt2

**Comment by yidinghabana** [comment]
Clean Unittests.
Delete AssertIsFile function.
Modify Test 6 with Subprocess.

File: ft-fine-tune-bert/fine-tune-unittests.py [hunk 12]
```diff
@@ -1,78 +1,19 @@
...
-    #         for row in self.eval_data:
-    #             for col in row:
-    #                 output.write(str(col) + '\n')
-
-
-    def assertIsFile(self, path):
-        if not pl.Path(path).resolve().is_file():
-            raise AssertionError("File does not exist: %s" % str(path))
-    
     def assertFileExist(self, path):
```
Comment (orionsBeltWest) [inline ft-fine-tune-bert/fine-tune-unittests.py]: Remove as not used.  (assetDirExist and FileExist)

File: ft-fine-tune-bert/fine-tune-unittests.py [hunk 13]
```diff
@@ -91,13 +32,14 @@ def assertDirExist(self, path):
     def test_data_paths(self):
         """Checks if input file existing"""
         file = pl.Path("./6_genre_clean_training_data_small.txt")
```
Comment (orionsBeltWest) [inline ft-fine-tune-bert/fine-tune-unittests.py]: File relative path as discussed:

Consider the following code snippet : 
```python:
        cfg_path = os.path.dirname(os.path.abspath(__file__))
        cfg_file = cfg_path + "/" + "connector_config.yaml"
```

File: ft-fine-tune-bert/fine-tune-unittests.py [hunk 14]
```diff
@@ -127,110 +69,27 @@ def test_return_type(self):
...
-            trainer.train()
-
-        # Evaluation
-        if training_args.do_eval:
-            logger.info("*** Evaluate ***")
-            eval_output = trainer.evaluate()
-            perplexity = math.exp(eval_output["eval_loss"])
+        bert_target = 220
+        
+        cmd = "python -u ./finetune.py --model_name bert-base-uncased "
```
Comment (orionsBeltWest) [inline ft-fine-tune-bert/fine-tune-unittests.py]: fix the path per comment above.  

Also recommended to not hard code these parameters.  In addition try to pass the training data files  and not assume the default?

**Review by orionsBeltWest — CHANGES_REQUESTED** [review]
Still few small but important changes:

1) Fix path as in comment. 
2) delete not used functions. 
3) Do not hard code test param. load from config file.

**Comment by yidinghabana** [comment]
No problem. I re-submit the codes for review by incorporating the following comments:

Clean unittests relative paths
Remove not using functions (assertDirExist and assertFileExist)
Parse inputs files in subprocess cmd

File: ft-fine-tune-distilgpt2/fine-tune-unittests.py [hunk 15]
```diff
@@ -75,16 +66,20 @@ def test_return_metrics(self):
         model_name = "distilgpt2"
         output_model_path = "story_generator_checkpoint_"
         distilgpt2_target = 90
-        
-        cmd = "python -u ./finetune.py --model_name distilgpt2 "
+        finetune_path = os.path.join(scripts_dir, 'finetune.py')
+
+        cmd = "python -u " + finetune_path + " --model_name distilgpt2 " \
```
Comment (orionsBeltWest) [inline ft-fine-tune-distilgpt2/fine-tune-unittests.py:77]: I do not see a setup function in which you are reading test parameters? It looks you are still hard coding this test params! 

What about the files path fix I mentioned in my previous review.  Is this implemented some place?

**Review by orionsBeltWest — CHANGES_REQUESTED** [review]

File: ft-fine-tune-distilgpt2/fine-tune-unittests.py (hunk 15 above)
Comment (yidinghabana) [reply ft-fine-tune-distilgpt2/fine-tune-unittests.py:77]: Basem, Thanks for your comments. I re-submitted the code for your review by parsing testing parameters from unittests as well. Originally I parsing testing parameters in main script by using library yamls. 

In addition, I fixed paths by using absolute paths in each function as below: 
scripts_dir = pl.Path(__file__).parent.resolve()
finetune_path = os.path.join(scripts_dir, 'finetune.py')

**Review by orionsBeltWest — APPROVED** [review]
looks good
